### PR TITLE
[SPARK-8603][SPARKR] Use shell() instead of system2() for SparkR on Windows

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

This PR corrects SparkR to use `shell()` instead of `system2()` on Windows.

Using `system2(...)` on Windows does not process windows file separator `\`. `shell(tralsate = TRUE, ...)` can treat this problem. So, this was changed to be chosen according to OS.

Existing tests were failed on Windows due to this problem. For example, those were failed.

```
8. Failure: sparkJars tag in SparkContext (@test_includeJAR.R#34)
9. Failure: sparkJars tag in SparkContext (@test_includeJAR.R#36)
```

The cases above were due to using of `system2`.

In addition, this PR also fixes some tests failed on Windows.

```
5. Failure: sparkJars sparkPackages as comma-separated strings (@test_context.R#128) 
6. Failure: sparkJars sparkPackages as comma-separated strings (@test_context.R#131) 
7. Failure: sparkJars sparkPackages as comma-separated strings (@test_context.R#134) 
```

  The cases above were due to a weird behaviour of `normalizePath()`. On Linux, if the path does not exist, it just prints out the input but it prints out including the current path on Windows.

``` r
# On Linus
path <- normalizePath("aa")
print(path)
[1] "aa"

# On Windows
path <- normalizePath("aa")
print(path)
[1] "C:\\Users\\aa"
```
## How was this patch tested?

Jenkins tests and manually tested in a Window machine as below:

Here is the [stdout](https://gist.github.com/HyukjinKwon/4bf35184f3a30f3bce987a58ec2bbbab) of testing.

Closes #7025

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Please let me cc @sun-rui and @JoshRosen

**Comment by sun-rui** [comment]
Could you add a test case for Windows specific test? @shivaram

File: R/pkg/R/client.R [hunk 1]
```diff
@@ -43,6 +43,17 @@ determineSparkSubmitBin <- function() {
   sparkSubmitBinName
 }
 
+# R supports both file separator in the file path (Unix : / and Windows: \) irrespective
+# of the operating system. So, this should be manaully changed.
+determinefileSeperator <- function() {
+  if (.Platform$OS.type == "windows") {
+    fileSeperator <- "\\"
```
Comment (srowen) [inline R/pkg/R/client.R]: Nit: separator not seperator

Comment (HyukjinKwon) [inline R/pkg/R/client.R]: Thank you so much.

**Comment by felixcheung** [comment]
Does this apply to other cases:
https://github.com/apache/spark/blob/d6dc12ef0146ae409834c78737c116050961f350/R/pkg/inst/worker/daemon.R#L22
https://github.com/apache/spark/blob/d6dc12ef0146ae409834c78737c116050961f350/R/pkg/inst/profile/shell.R#L20
https://github.com/apache/spark/blob/6ab4d9e0c76b69b4d6d5f39037a77bdfb042be19/examples/src/main/r/dataframe.R#L37

(last one is an example)

**Comment by HyukjinKwon** [comment]
@sun-rui @felixcheung Let me try to build and run all tests for R first in Windows and then will try to correct and add each test one by one. This will take a bit of time and I might have to ask a lot of questions but anyway I will try.

**Comment by sun-rui** [comment]
@felixcheung, this issue seems to relate to system2() only. However, let's wait for HyukjinKwon's test result.

@HyukjinKwon, great, go ahead please.

**Comment by HyukjinKwon** [comment]
@sun-rui @felixcheung Right. It seems finally I made it. I made gists and upload a PDF file for Spark UI.

Let me tell you the test results first.

Here is the stdout output for the tests on Windwos 7 32bit, [output.msg](https://gist.github.com/HyukjinKwon/6a10719d2ca67e04ece2b23a8f92dc62).

Here is the stderr output for the tests on Windwos 7 32bit, [output.err](https://gist.github.com/HyukjinKwon/54984d57ee18236d46e965d07b31f77a).

Here is the PDF for [Spark UI PDF](https://drive.google.com/open?id=0B7RfLjRU7QTnVVA2bkVMVFkzNEE)

1. I run tests after building Spark on Windows according to [`./R/WINDOWS.md`] (https://github.com/apache/spark/blob/master/R/WINDOWS.md)

2. It seems `$HADOOP_HOME` should be set.

3. It seems `winutils.exe` is required which is included in Hadoop official binary although it reads file in the local file system.

4. And run the tests by the command below:

```bash
cd bin
spark-submit2.cmd --conf spark.hadoop.fs.defualt.name="file:///" ..\R\pkg\tests\run-all.R > output.msg 2> output.err
```

**Comment by HyukjinKwon** [comment]
This raises some questions to me.

1. It seems several tests were failed. Could you please inform me your thoughts?

2. Now, I think I can add some tests but could you please where I should write the related tests and maybe rough ideas of the tests I should add?

**Comment by sun-rui** [comment]
1. A rough scan of the test failures shows most of them are probably related to path handling. You can replay the failed test case in R on Windows.  For debug facilities in R, refer to http://www.inside-r.org/r-doc/base/traceback, https://stat.ethz.ch/R-manual/R-devel/library/base/html/debug.html, http://www.inside-r.org/r-doc/base/browser
2. You can add a new file named test_Windows.R under R/pkg/inst/tests/testthat
3. Sure.

**Comment by sun-rui** [comment]
I checked the source code of system2. Since a batch file with .cmd extension is to be launched on Windows, I think it is more formal to use shell() instead of system2(). Refer to http://www.astrostatistics.psu.edu/datasets/R/html/base/html/shell.html. The proposed fix is like follows:
```
if (.Platform$OS.type == "windows") {
  shell(sparkSubmitBin, translate = TRUE, ...)
} else {
  system2(sparkSubmitBin, ...)
}
```

**Comment by HyukjinKwon** [comment]
@sun-rui Thank you so much. I will try to add a test.

**Comment by shivaram** [comment]
Related to my comment on #13217 -- I will test this out on windows using the new instructions.

**Comment by HyukjinKwon** [comment]
@shivaram Thank you so much. Let me try to add a test meanwhile.

**Comment by HyukjinKwon** [comment]
@shivaram and @sun-rui I addressed the comment, added a test for this PR and fixed some tests not working on Windows due to path separator.

This PR also fixes the test failures below.

- **Fixed case 1**

  ```
5. Failure: sparkJars sparkPackages as comma-separated strings (@test_context.R#128) 
6. Failure: sparkJars sparkPackages as comma-separated strings (@test_context.R#131) 
7. Failure: sparkJars sparkPackages as comma-separated strings (@test_context.R#134) 
```

  The cases above were due to a weird behaviour of `normalizePath()`. On Linux, if the path does not exist, it just prints out the input but it prints out including the current path on Windows.

  ```r
# On Linus
path <- normalizePath("aa")
print(path)
[1] "aa"

# On Windows
path <- normalizePath("aa")
print(path)
[1] "C:\\Users\\aa"
```

- **Fixed case 2**

  ```
8. Failure: sparkJars tag in SparkContext (@test_includeJAR.R#34)
9. Failure: sparkJars tag in SparkContext (@test_includeJAR.R#36)
```

The cases above were due to using of `system2`.

Here is full stdout of my testing, [output.msg](https://gist.github.com/HyukjinKwon/c72b747a15948bd062a26aba7c42115b) on Windows.

File: R/pkg/R/client.R [hunk 2]
```diff
@@ -60,6 +60,15 @@ generateSparkSubmitArgs <- function(args, sparkHome, jars, sparkSubmitOpts, pack
   combinedArgs
 }
 
+determineLauncher <- function(sparkSubmitBin, combinedArgs, capture = FALSE) {
```
Comment (sun-rui) [inline R/pkg/R/client.R]: can this be a common utility function and be re-used in test_includeJAR.R?

Comment (HyukjinKwon) [inline R/pkg/R/client.R]: Sure. I will. 

BTW, could you inform me if I can disable style checking for some blocks? It seems it fails due to the usages of `shell`.

Comment (shivaram) [inline R/pkg/R/client.R]: You can use `nolint` - See `context.R` for an example. BTW what is the lint error here ?

Comment (HyukjinKwon) [inline R/pkg/R/client.R]: @shivaram For some reasons, it seems there is no mothod `shell()` on Linux (I tested this on my Mac) but on Windows. Here is a thread for this, http://r.789695.n4.nabble.com/Shell-Function-not-on-Linux-td4672569.html.

I could not find the official documentation though. Let me address the comments first.

Comment (sun-rui) [inline R/pkg/R/client.R]: the common function could be like:
```
  launchScript(script) {
    if (windowsOS) use shell()
    else (system2()
  }
```
This makes it more general, not binding to launching SparkSubmit

Comment (HyukjinKwon) [inline R/pkg/R/client.R]: @sun-rui Thank you so much. I was trying to come up with a good name.

**Comment by HyukjinKwon** [comment]
I just addressed the comments and here is the (stdout)[https://gist.github.com/HyukjinKwon/4bf35184f3a30f3bce987a58ec2bbbab] of testing on Windiws.

**Comment by HyukjinKwon** [comment]
retest this please

File: R/pkg/inst/tests/testthat/test_includeJAR.R [hunk 3]
```diff
@@ -21,10 +21,13 @@ runScript <- function() {
   sparkTestJarPath <- "R/lib/SparkR/test_support/sparktestjar_2.10-1.0.jar"
   jarPath <- paste("--jars", shQuote(file.path(sparkHome, sparkTestJarPath)))
   scriptPath <- file.path(sparkHome, "R/lib/SparkR/tests/testthat/jarTest.R")
-  submitPath <- file.path(sparkHome, "bin/spark-submit")
-  res <- system2(command = submitPath,
-                 args = c(jarPath, scriptPath),
-                 stdout = TRUE)
+  if (.Platform$OS.type == "windows") {
```
Comment (sun-rui) [inline R/pkg/inst/tests/testthat/test_includeJAR.R]: you can call determineSparkSubmitBin() here

**Comment by HyukjinKwon** [comment]
@sun-rui Thanks, I just updated it. There is the [stdout](https://gist.github.com/HyukjinKwon/c34f7b5581de46433a4981f1d30b0421) for this time

**Comment by sun-rui** [comment]
LGTM

**Comment by sun-rui** [comment]
could you rebase it to latest master. It is hopeful other test failures will gone

**Comment by HyukjinKwon** [comment]
Sure, Thanks!

**Comment by shivaram** [comment]
@HyukjinKwon Thanks for the update. I verified this today using a Windows VM and followed the instructions you have written up as well. All the tests passed and the stdout I got is at https://gist.github.com/shivaram/bd3c8bf7466d6cd515f4877f2a9611b3

One minor thing - there is a typo in the last line of https://github.com/apache/spark/blob/branch-2.0/R/WINDOWS.md
`spark.hadoop.fs.defualt.name` should be `spark.hadoop.fs.default.name` (typo in the spelling of `default`). 

Could you make that change as a part of this PR ?

**Comment by HyukjinKwon** [comment]
@shivaram Thank you so much! I will keep this machine as it is anyway. Please cc me if there are some PRs dealing with Windows in the future. I can run tests and leave some comments.

**Comment by sun-rui** [comment]
@shivaram

**Comment by shivaram** [comment]
Merging this to master and branch-2.0

**Pull request closed without merging**